### PR TITLE
RPC-126-separate rtm

### PR DESCRIPTION
**Pull request by SKonstantinS**
Rarible transfer manager now is separate contract

### CODE REVIEW

**Comment by disenotov** [comment]
why do we change domain name to ExchangeMetaV2 now?

File: exchange-v2/contracts/ExchangeV2Core.sol [hunk 1]
```diff
@@ -3,27 +3,32 @@
...
 
     //state of the orders
     mapping(bytes32 => uint) public fills;
-
+    ITransferManager public raribleTransferManager;
     //events
     event Cancel(bytes32 hash, address maker, LibAsset.AssetType makeAssetType, LibAsset.AssetType takeAssetType);
     event Match(bytes32 leftHash, bytes32 rightHash, address leftMaker, address rightMaker, uint newLeftFill, uint newRightFill, LibAsset.AssetType leftAsset, LibAsset.AssetType rightAsset);
 
+    function setTransferManager(ITransferManager newRaribleTransferManager) external {
```
[the remark this answers is not in the document]
Comment (disenotov) [reply exchange-v2/contracts/ExchangeV2Core.sol:28]: i mean now there should be probably EchangeV2Core_init that calls this function, and __ExchangeV2_init calls EchangeV2Core_init

File: exchange-v2/contracts/ExchangeV2Core.sol [hunk 2]
```diff
@@ -51,54 +56,67 @@ abstract contract ExchangeV2Core is Initializable, OwnableUpgradeable, AssetMatc
...
-        LibOrderDataV2.DataV2 memory rightOrderData = LibOrderData.parse(orderRight);
+        (LibDeal.Data memory leftDealData, bool fillLeft) = LibOrderData.parse(orderLeft);
+        (LibDeal.Data memory rightDealData, bool fillRight) = LibOrderData.parse(orderRight);
 
-        LibFill.FillResult memory newFill = getFillSetNew(orderLeft, orderRight, leftOrderKeyHash, rightOrderKeyHash, leftOrderData, rightOrderData);
+        LibFill.FillResult memory newFill = getFillSetNew(orderLeft, orderRight, LibOrder.hashKey(orderLeft), LibOrder.hashKey(orderRight), fillLeft, fillRight);
 
-        (uint totalMakeValue, uint totalTakeValue) = doTransfers(makeMatch, takeMatch, newFill, orderLeft, orderRight, leftOrderData, rightOrderData);
+        runTransfers(makeMatch, takeMatch, newFill, leftDealData, rightDealData, orderLeft.maker, orderRight.maker);
+        emit Match(LibOrder.hashKey(orderLeft), LibOrder.hashKey(orderRight), orderLeft.maker, orderRight.maker, newFill.rightValue, newFill.leftValue, makeMatch, takeMatch);
```
Comment (disenotov) [inline exchange-v2/contracts/ExchangeV2Core.sol:66]: we already parsed those orders for data on line 63, so it's better to store those in variables to not parse it twice

File: exchange-v2/contracts/ExchangeV2Core.sol [hunk 3]
```diff
@@ -51,54 +56,67 @@ abstract contract ExchangeV2Core is Initializable, OwnableUpgradeable, AssetMatc
...
+            address(raribleTransferManager).transferEth(msg.value);
         }
-        emit Match(leftOrderKeyHash, rightOrderKeyHash, orderLeft.maker, orderRight.maker, newFill.rightValue, newFill.leftValue, makeMatch, takeMatch);
+        LibAsset.Asset memory makeMatchAsset;
+        LibAsset.Asset memory takeMatchAsset;
+        makeMatchAsset.assetType = makeMatch;
+        makeMatchAsset.value = newFill.leftValue;
+        takeMatchAsset.assetType = takeMatch;
+        takeMatchAsset.value = newFill.rightValue;
+        ITransferManager(raribleTransferManager).doTransfers(makeMatchAsset, takeMatchAsset, leftDealData, rightDealData, leftMaker, rightMaker, _msgSender());
```
Comment (disenotov) [inline exchange-v2/contracts/ExchangeV2Core.sol:93]: we could probably do something like doTransfers(LibAsset.Asset(makeMatch, newFill.leftValue))... if it saves gas
i mean send it without creating extra variables

File: exchange-v2/test/contracts/v2/SimpleTransferManager.sol [hunk 4]
```diff
@@ -1,28 +1,43 @@
...
 pragma solidity 0.7.6;
+pragma abicoder v2;
 
-import "../../../contracts/ITransferManager.sol";
+import "@rarible/exchange-interfaces/contracts/ITransferManager.sol";
+import "@rarible/transfer-manager/contracts/TransferExecutor.sol";
 
-abstract contract SimpleTransferManager is ITransferManager {
-    using SafeMathUpgradeable for uint;
+contract SimpleTransferManager is TransferExecutor, ITransferManager {
```
Comment (disenotov) [inline exchange-v2/test/contracts/v2/SimpleTransferManager.sol:9]: there is a warning about receive() instead fallback(), need to decide what we need to use here (i found something here https://stackoverflow.com/questions/59651032/why-does-solidity-suggest-me-to-implement-a-receive-ether-function-when-i-have-a)

File: transfer-manager/contracts/RaribleTransferManager.sol [hunk 5]
```diff
@@ -60,39 +69,42 @@ abstract contract RaribleTransferManager is OwnableUpgradeable, ITransferManager
...
         } else {
-            transferPayouts(makeMatch, fill.leftValue, leftOrder.maker, rightOrderData.payouts, TO_TAKER);
-            transferPayouts(takeMatch, fill.rightValue, rightOrder.maker, leftOrderData.payouts, TO_MAKER);
+            transferPayouts(makeMatch.assetType, makeMatch.value, leftMaker, right.payouts, TO_TAKER);
+            transferPayouts(takeMatch.assetType, takeMatch.value, rightMaker, left.payouts, TO_MAKER);
+        }
+        /*Return back eth, if eth more than need*/
+        uint ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            address(originalMessageSender).transferEth(ethBalance);
```
Comment (disenotov) [inline transfer-manager/contracts/RaribleTransferManager.sol:94]: not sure if transferring all balance is ok

File: exchange-v2/test/contracts/v2/SimpleTransferManager.sol (hunk 4 above)
Comment (disenotov) [reply exchange-v2/test/contracts/v2/SimpleTransferManager.sol:9]: it's more about real raribleTransferManager but it concernes this too

**Review by disenotov — CHANGES_REQUESTED** [review]

File: exchange-v2/contracts/EmptyGap.sol [hunk 6]
```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.7.6;
+
+/*@dev for save context*/
+abstract contract EmptyGap {
+      function emptyDo() public view returns(uint) {
```
[the remark this answers is not in the document]
Comment (SKonstantinS) [reply exchange-v2/contracts/EmptyGap.sol:7]: delete, not need

File: exchange-v2/contracts/ExchangeMetaV2.sol [hunk 7]
```diff
@@ -4,24 +4,16 @@ pragma solidity 0.7.6;
...
-        address newDefaultFeeReceiver,
-        IRoyaltiesProvider newRoyaltiesProvider
     ) external initializer {
         __Context_init_unchained();
         __Ownable_init_unchained();
-        __TransferExecutor_init_unchained(_transferProxy, _erc20TransferProxy);
-        __RaribleTransferManager_init_unchained(newProtocolFee, newDefaultFeeReceiver, newRoyaltiesProvider);
         __OrderValidator_init_unchained();
-        __MetaTransaction_init_unchained("ExchangeV2", "1");
+        __MetaTransaction_init_unchained("ExchangeMetaV2", "1");
```
[the remark this answers is not in the document]
Comment (SKonstantinS) [reply exchange-v2/contracts/ExchangeMetaV2.sol:16]: ExchangeMetaV2 - its contract name, use the same for  domain name

File: exchange-v2/contracts/ExchangeV2Core.sol [hunk 8]
```diff
@@ -3,27 +3,32 @@
...
+abstract contract ExchangeV2Core is Initializable, OwnableUpgradeable, AssetMatcher, EmptyGap2, OrderValidator {
     using SafeMathUpgradeable for uint;
     using LibTransfer for address;
 
     uint256 private constant UINT256_MAX = 2 ** 256 - 1;
 
     //state of the orders
     mapping(bytes32 => uint) public fills;
-
+    ITransferManager public raribleTransferManager;
```
[the remark this answers is not in the document]
Comment (SKonstantinS) [reply exchange-v2/contracts/ExchangeV2Core.sol:23]: Error, fix

File: libraries/contracts/LibOrderData.sol [hunk 9]
```diff
@@ -3,22 +3,26 @@
 pragma solidity 0.7.6;
 
 import "./LibOrder.sol";
+import "./LibDeal.sol";
 
 library LibOrderData {
-    function parse(LibOrder.Order memory order) pure internal returns (LibOrderDataV2.DataV2 memory dataOrder) {
```
[the remark this answers is not in the document]
Comment (SKonstantinS) [reply libraries/contracts/LibOrderData.sol:8]: LibDeal need for ITransferManager
LibDeal combine all data about payouts and originFees
ITransferManager now don`t know about LibOrderDataV1 or LibOrderDataV2

File: exchange-v2/contracts/ExchangeV2Core.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (SKonstantinS) [reply exchange-v2/contracts/ExchangeV2Core.sol]: agree, done

File: exchange-v2/contracts/ExchangeV2Core.sol (hunk 2 above)
Comment (SKonstantinS) [reply exchange-v2/contracts/ExchangeV2Core.sol]: agree, done

File: exchange-v2/contracts/ExchangeV2Core.sol (hunk 3 above)
Comment (SKonstantinS) [reply exchange-v2/contracts/ExchangeV2Core.sol]: agree, done

File: exchange-v2/test/contracts/v2/SimpleTransferManager.sol (hunk 4 above)
Comment (SKonstantinS) [reply exchange-v2/test/contracts/v2/SimpleTransferManager.sol:9]: agree, use receive() instead fallback()

File: transfer-manager/contracts/RaribleTransferManager.sol (hunk 5 above)
Comment (SKonstantinS) [reply transfer-manager/contracts/RaribleTransferManager.sol]: eah, send eth in call doTransfers()
inside  doTransfers() get eth by msg.value

**Comment by SKonstantinS** [comment]
close, because open  Rpc-126-separate-RT- plus-auctions #119 , separate RTM in it

**Pull request closed without merging**